### PR TITLE
feat(common/models): debug-mode predictive-text worker sourcemapping 🧩

### PR DESCRIPTION
**Pull request by jahorton**
This PR introduces the compilation steps + wiring necessary to facilitate attaching sourcemaps to the worker's source blob.  When enabled, this will provide sourcemaps within the worker thread, which can greatly facilitate debugging therein!

Note:  part of this resulted in #7894 - commit https://github.com/keymanapp/keyman/pull/8052/commits/42531aed233bfb76698fe3a46e3e28074578619f should actually be remarkably similar.  I minorly bungled up the rebases involved and excluded that PR by mistake.  (Note:  we're **8 PRs deep** on the rebase here!)  Still, it should be "simple enough" for me to resolve any merge conflicts when the time comes; it was mostly complete when I spun it off.

### CODE REVIEW

**Comment by jahorton** [comment]
Note:  as of the last commit...

image

From here out (once changes are merged into descendants), this unit-test set should pass.

**Comment by jahorton** [comment]
Possibly worth note: https://esbuild.github.io/api/#define may give us the avenue needed here.  I haven't experimented with that yet.

**Comment by jahorton** [comment]
... and, about half a day's experimentation later... I've got something.

The worker code may now be _both_ minified _and_ sourcemapped - before this, it'd actually been a "one or the other" sort of thing, since the encoding setup used to ensure proper mapping stability doesn't minify well.  Minification is now applied as part of the worker-wrapping setup, before its source is encoded into the final wrapped version.  (It's being done with `esbuild`.)

The scripts in `common/web/lm-worker` now have appropriate options (`--debug`, `--minify`) that may be set to modify the build output:
- `--debug`:  the worker's sourcemaps will include the original files' source contents.  If not set, this will be erased from the map, preventing heavy filesize bloat.
    - I have opted to keep the sourcemaps attached in this case; https://esbuild.github.io/api/#sources-content suggests that we should still be able to get unminified stack traces from Sentry tooling even without including the `sourcesContents` sourcemap property.
        - Of course, this has yet to be verified.
    - `--minify` and `--debug`:   the wrapped worker code clocks in ~726 KB.
    - `--minify`, no `--debug`:  291 KB.
        - Actual script, minified:  94 KB.
        - Sourcemaps _without `sourcesContents`_: 120 KB.
        - So... we may wish to actually full-on drop them then.

Note that there's still more tooling to do for the `common/predictive-text` script (or a higher-level one) to be able to select which version to use.  That said... esbuild offers this tool, which may help us down the line:  https://esbuild.github.io/api/#inject

**Comment by jahorton** [comment]
OK... took me a while to sort through the possibilities, and I think I've found the best way forward given what is facilitated by the toolchain.

Long story short, `lm-worker` builds now produce two artifacts:  a non-minified version with full sourcemaps and a minified version without sourcemaps.  (We can add a third if we wish to / need to for Sentry error report analysis.)  These are made available as TS modules.

Note that #8053's changes (already, fortunately) include the need to specify a `Worker` instance to use in its constructor.  Generally, a top-level build product will choose a single type of `Worker` to supply to `LanguageProcessor` instances; any unused ones supplied by the lower-level modules will be dropped by bundler 'tree-shaking' optimizations.

Just... took me a while to start fully "thinking in modules".

File: common/predictive-text/unit_tests/in_browser/cases/worker.js [hunk 1]
```diff
@@ -10,9 +10,10 @@ describe('LMLayerWorker', function () {
 
   describe('LMLayerWorkerCode', function() {
     it('should exist!', function() {
-      assert.isFunction(LMLayerWorkerCode,
-        'Could not find LMLayerWorkerCode! Does embedded_worker.js exist?'
-      );
+      // assert.isFunction(LMLayerWorkerCode,
+      //   'Could not find LMLayerWorkerCode! Does embedded_worker.js exist?'
+      // );
+      assert.isString(LMLayerWorkerCode);
```
Comment (mcdurdin) [inline common/predictive-text/unit_tests/in_browser/cases/worker.js:16]: You've left code commented here. What is the long-term intention?

File: common/tools/sourcemap-path-remapper/build.sh [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+## START STANDARD BUILD SCRIPT INCLUDE
+# adjust relative paths as necessary
+THIS_SCRIPT="$(greadlink -f "${BASH_SOURCE[0]}" 2>/dev/null || readlink -f "${BASH_SOURCE[0]}")"
+. "$(dirname "$THIS_SCRIPT")/../../../resources/build/build-utils.sh"
+## END STANDARD BUILD SCRIPT INCLUDE
+
+. "$KEYMAN_ROOT/resources/shellHelperFunctions.sh"
+
+# This script runs from its own folder
+cd "$(dirname "$THIS_SCRIPT")"
```
Comment (mcdurdin) [inline common/tools/sourcemap-path-remapper/build.sh:21]:
```suggestion
cd "$THIS_SCRIPT_PATH"
```

File: common/tools/sourcemap-path-remapper/build.sh [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  configure clean build
+
+builder_describe_outputs \
+  configure     /node_modules \
+  build         build/index.js
+
+builder_parse "$@"
+
+# TODO: build if out-of-date if test is specified
+# TODO: configure if npm has not been run, and build is specified
```
Comment (mcdurdin) [inline common/tools/sourcemap-path-remapper/build.sh:40]: I think these two comments are no longer needed?

```suggestion
```

File: common/tools/sourcemap-path-remapper/build.sh [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+if builder_start_action configure; then
+  verify_npm_setup
+  builder_finish_action success configure
+fi
+
+if builder_start_action clean; then
+  npm run clean
+  builder_finish_action success clean
+fi
```
Comment (mcdurdin) [inline common/tools/sourcemap-path-remapper/build.sh:50]: Normally a `clean` would precede a `configure`?

File: common/tools/sourcemap-path-remapper/package.json [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  "main": "build/index.js",
+  "type": "module",
+  "scripts": {
+    "tsc": "tsc"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/keymanapp/keyman.git"
+  },
+  "author": "Joshua A. Horton <jahorton@sil.org>",
```
Comment (mcdurdin) [inline common/tools/sourcemap-path-remapper/package.json:13]:
```suggestion
  "author": "Joshua A. Horton <joshua_horton@sil.org>",
```

File: common/tools/sourcemap-path-remapper/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,139 @@
+// Sourcemap types!
+
+import * as fs from 'fs';
+import convertSourcemap from 'convert-source-map'; // Transforms sourcemaps among various common formats.
```
Comment (mcdurdin) [inline common/tools/sourcemap-path-remapper/src/index.ts:4]: ~~I was mad at someone else but now I am mad at you :-P~~

```suggestion
import convertSourceMap from 'convert-source-map'; // Transforms sourcemaps among various common formats.
```

File: common/tools/sourcemap-path-remapper/src/index.ts [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+ * mutate it.
+ */
+class RemappingState {
+  constructor(srcmap: Object) {
+    this.sourcemap = srcmap;
+    this.resetChangedSourcepathTracking();
+  }
+
+  public readonly sourcemap: any;
+  private _unchangedSourcepaths: string[] = [];
```
Comment (mcdurdin) [inline common/tools/sourcemap-path-remapper/src/index.ts:26]:
```suggestion
  public readonly sourceMap: any;
  private _unchangedSourcePaths: string[] = [];
```

and so on...

File: common/web/lm-worker/build-polyfill-concatenator.js [hunk 8]
```diff
@@ -0,0 +1,147 @@
+import fs from 'fs';
+
+import SourcemapRemapper from '@keymanapp/sourcemap-path-remapper';
+
+import SourcemapCombiner from 'combine-source-map';
+import convertSourcemap from 'convert-source-map'; // Transforms sourcemaps among various common formats.
```
Comment (mcdurdin) [inline common/web/lm-worker/build-polyfill-concatenator.js:6]:
```suggestion
import SourceMapRemapper from '@keymanapp/sourcemap-path-remapper';

import SourceMapCombiner from 'combine-source-map';
import convertSourceMap from 'convert-source-map'; // Transforms sourcemaps among various common formats.
```

File: common/web/lm-worker/build-polyfill-concatenator.js [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+                                                   // Base64, stringified-JSON, end-of-file comment...
+
+let loadPolyfill = function(file, mapFilename) {
```
Comment (mcdurdin) [inline common/web/lm-worker/build-polyfill-concatenator.js:9]:
```
let loadPolyfill = function(file, mapFileName) {
```

OK NOT REALLY BUT JUST FOR FUN

File: common/web/lm-worker/src/main/index.ts [hunk 10]
```diff
@@ -30,7 +30,7 @@
  */
 
 /// <reference types="@keymanapp/lm-message-types" />
-import type * as types from "@keymanapp/web-utils/build/obj/kmwstring.js";
+import { extendString } from "@keymanapp/web-utils/build/obj/index.js";
```
Comment (mcdurdin) [inline common/web/lm-worker/src/main/index.ts:33]: `extendString` does not appear to be used in this file?

File: common/web/lm-worker/src/main/worker-main.ts [hunk 11]
```diff
@@ -21,7 +21,7 @@
  */
 
 import LMLayerWorker from './index.js';
-export { default as LMLayerWorker } from './index.js';
+//export { default as LMLayerWorker } from './index.js';
```
Comment (mcdurdin) [inline common/web/lm-worker/src/main/worker-main.ts:24]: Probably should be deleted?
```suggestion
```

**Review by mcdurdin — APPROVED** [review]
Minor changes requested; `sourceMap` vs `sourcemap` changes entirely up to you.

**Pull request merged**